src/arch/x86_64/CodeGen.zig

```diff
@@ -3772,13 +3799,17 @@ fn store(self: *Self, ptr: MCValue, value: MCValue, ptr_ty: Type, value_ty: Type
             const addr_reg_lock = self.register_manager.lockRegAssumeUnused(addr_reg);
             defer self.register_manager.unlockReg(addr_reg_lock);
 
-            try self.loadMemPtrIntoRegister(addr_reg, ptr_ty, ptr);
-            // Load the pointer, which is stored in memory
-            try self.asmRegisterMemory(
-                .mov,
-                addr_reg.to64(),
-                Memory.sib(.qword, .{ .base = addr_reg.to64() }),
-            );
+            if (ptr.isThreadlocal()) {
+                try self.loadThreadLocalPtrIntoRegister(addr_reg, ptr_ty, ptr);
+            } else {
+                try self.loadMemPtrIntoRegister(addr_reg, ptr_ty, ptr);
+                // Load the pointer, which is stored in memory
+                try self.asmRegisterMemory(
+                    .mov,
+                    addr_reg.to64(),
+                    Memory.sib(.qword, .{ .base = addr_reg.to64() }),
+                );
```

Comment:
Why would this not apply to thread local pointers?  `ptr` is a global pointer, so once you have a pointer to the global, you have to load the address out of memory.  Any instance where the `isThreadLocal` check can't be moved into `loadMemPtrIntoRegister` seems suspect.

Reply:
We end up doing a double dereference at the wrong time. For standard GOT deref we do something like:

```
lea reg, [rip]
mov reg, [reg]
mov value_reg, [reg]
```

while for TLV we don't want the middle deref as we get the address of a thread-local variable:

```
mov rdi, [rip]
call [rdi]
mov reg, rax
mov value_reg, [reg]
```